src/tactic/linear_combination.lean

```diff
@@ -0,0 +1,392 @@
+`normalize` field of the given configuration is true.
+
+* Input:
+  * `config` : a `linear_combination_config`, which determines the tactic used
+      for normalization if normalization is done
+
+* Output: N/A
+-/
+meta def prove_equal_if_desired (config : linear_combination_config) :
+  tactic unit :=
+when config.normalize config.normalization_tactic
+
+/-! ### Part 4: Completed Tactic -/
+
+
+/--
+This is a tactic that attempts to prove the target by creating and applying a
+  linear combination of a list of equalities.  (If the `normalize` field of the
+  configuration is set to ff, then the tactic will simply set the user up to
+  prove their target using the linear combination instead of attempting to
+  finish the proof.)
+
+Note: The left and right sides of all the equalities should have the same
+  ring type, and the coefficients should also have this type.  There must be
+  instances of `has_mul` and `add_group` for this type.  Also note that the
+  target must involve at least one variable.
+
+* Input:
+  * `h_eqs_names` : a list of names, referring to equations in the local
+      context
+  * `coeffs` : a list of coefficients to be multiplied with the corresponding
+    equations in the list of names
+  * `config` : a `linear_combination_config`, which determines the tactic used
+      for normalization; by default, this value is the standard configuration
+      for a `linear_combination_config`
+
+* Output: N/A
+-/
+meta def linear_combination (h_eqs_names : list name) (coeffs : list pexpr)
+  (config : linear_combination_config := {}) : tactic unit :=
+do
+  hsum ← make_sum_of_hyps h_eqs_names coeffs,
+  hsum_on_left ← move_to_left_side hsum,
+  move_target_to_left_side,
+  set_goal_to_hleft_eq_tleft hsum_on_left,
+  prove_equal_if_desired config
+
+
+section interactive_mode
+setup_tactic_parser
+
+/--
+A parser that matches a pair in parentheses, where the first item in the pair
+is an identifier and the second item in the pair is a `pexpr`.
+
+* Input: None
+
+* Output: a `lean.parser (name × pexpr)`
+-/
+meta def parse_name_pexpr_pair : lean.parser (name × pexpr) :=
+do
+  tk "(",
+  id ← ident,
+  tk ",",
+  coeff ← parser.pexpr 0,
+  tk ")",
+  pure (id, coeff)
+
+/--
+`linear_combination` attempts to prove the target by creating and applying a
+  linear combination of a list of equalities.  The tactic will create a linear
+  combination by adding the equalities together from left to right, so the order
+  of the input hypotheses does matter.  If the `normalize` field of the
+  configuration is set to false, then the tactic will simply set the user up to
+  prove their target using the linear combination instead of attempting to
+  finish the proof.
+
+Note: The left and right sides of all the equalities should have the same
+  type, and the coefficients should also have this type.  There must be
+  instances of `has_mul` and `add_group` for this type.  Also note that the
+  target must involve at least one variable.
```

Comment:
@agoldb10 could you explain what you mean by "the target must involve at least one variable"? I didn't see any failing tests that addressed this, and I'm not sure what it means, so I'm wondering if it needs to be mentioned at all!

Reply:
That is a great question.  I wrote that comment a while ago, but I think I believed the tactic does not work when the target equality or at least one of the given equalities does not contain any variables.  For example, the following two tests fail:

```
example (x y : ℤ) (h1 : x + 4 = 2) (h2 : -x = 2) :
  4 = 4 :=
by linear_combination (h1, 1) (h2, 1)

example (x y : ℤ) (h1 : 2 + x = 3) (h2 : 2 = 2) :
  2 + x + 2 = 3 + 2 :=
by linear_combination (h1, 1) (h2, 1)
```

However, looking at this now, it seems like the problem in those tests is actually that the lemma applications are missing the implicit arguments about the types of the terms in the equalities.  So, these problems can be fixed by just specifying the type of a term in the equality without a variable, as in the test below:
```
example (x y : ℤ) (h1 : x + 4 = 2) (h2 : -x = 2) :
  (4 : ℤ) = 4 :=
by linear_combination (h1, 1) (h2, 1)
```

Do you think this behavior should be described by modifying the comment and adding tests, or do you not think this is worth mentioning?

Reply:
Ah, I see! The problem is that `4 = 4` is by default an equality about `nat`, while the hypotheses are about `int`.

I think this is fine, and covered by the previous disclaimer, so you could delete this line.

Reply:
Sounds good, I just deleted the line.  Thank you for noticing this!